src/sinks/clickhouse.rs

```diff
@@ -0,0 +1,295 @@
+    } else {
+        format!("{}/?{}", host, query)
+    };
+
+    url.parse::<Uri>()
+        .map_err(|e| format!("Unable to parse host as URI: {}", e))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn encode_valid() {
+        let uri = encode_uri("http://localhost:80", "my_database", "my_table").unwrap();
+        assert_eq!(uri, "http://localhost:80/?query=INSERT+INTO+%22my_database%22.%22my_table%22+FORMAT+JSONEachRow");
+
+        let uri = encode_uri("http://localhost:80", "my_database", "my_\"table\"").unwrap();
+        assert_eq!(uri, "http://localhost:80/?query=INSERT+INTO+%22my_database%22.%22my_%5C%22table%5C%22%22+FORMAT+JSONEachRow");
+    }
+
+    #[test]
+    fn encode_invalid() {
+        encode_uri("localhost:80", "my_database", "my_table").unwrap_err();
+    }
+}
+
+#[cfg(test)]
+#[cfg(feature = "clickhouse-integration-tests")]
+mod integration_tests {
+    use super::*;
+    use crate::buffers::Acker;
+    use crate::{
+        test_util::{block_on, random_string},
+        topology::config::SinkConfig,
+        Event,
+    };
+    use futures::Sink;
+    use serde_json::Value;
+
+    #[test]
+    fn insert_events() {
+        let table = gen_table();
+        let host = String::from("http://localhost:8123");
+
+        let config = ClickhouseConfig {
+            host: host.clone(),
+            table: table.clone(),
+            compression: Some(Compression::None),
+            batch_size: Some(1),
+            request_retry_attempts: Some(1),
+            ..Default::default()
+        };
+
+        let client = ClickhouseClient::new(host);
+        client.create_table(&table);
+
+        let (sink, _hc) = config.build(Acker::Null).unwrap();
+
+        let mut input_event = Event::from("raw log line");
+        input_event
+            .as_mut_log()
+            .insert_explicit("host".into(), "example.com".into());
+
+        let pump = sink.send(input_event.clone());
+        block_on(pump).unwrap();
+
+        let output = client.select_all(&table);
+        assert_eq!(1, output.rows);
+
+        let expected = serde_json::to_value(input_event.into_log().all_fields()).unwrap();
+        assert_eq!(expected, output.data[0]);
+    }
+
+    struct ClickhouseClient {
+        host: String,
+        client: reqwest::Client,
+    }
+
+    impl ClickhouseClient {
+        fn new(host: String) -> Self {
+            ClickhouseClient {
+                host,
+                client: reqwest::Client::new(),
+            }
+        }
+
+        fn create_table(&self, table: &str) {
+            let mut response = self
+                .client
+                .post(&self.host)
+                .body(format!(
+                    "CREATE TABLE {}
+                     (host String, timestamp String, message String)
+                     ENGINE = MergeTree()
+                     PARTITION BY substring(timestamp, 1, 7)
+                     ORDER BY (host, timestamp);",
+                    table,
```

Comment:
As Vector controls serialization to JSON, it is possible to use real timestamps in ClickHouse: create table like

```sql
"CREATE TABLE {}
(host String, timestamp DateTime('UTC'), message String)
ENGINE = MergeTree()
PARTITION BY toYYYYMM(timestamp)
ORDER BY (host, timestamp);,
```

and then serialize timestamps to strings as `2019-08-07 07:01:25`. The reason ClickHouse doesn't understand RFC 3339 format is probably that timestamps in ClickHouse don't have sub-second fractions, so ClickHouse doesn't parse timestamps with sub-second fractions to not lose precision.

Reply:
Yeah, I'm not sure exactly how we can accomplish this via our serialization since the encoding to RFC3339 is done pretty deep within the serialization tree.

@lukesteensen is on vacation right now, but I'd like to bring this up with him once he gets back. 

We should probably open an issue around this and the load balancing ideas.

Reply:
Also `i64` numbers should probably be serialized as strings because JSON standard doesn't allow more than 53 bits of precision.

Reply:
There's probably some internal pre-serialization transform we could do for this that hopefully wouldn't be too complex. Definitely worth opening a followup issues for these things imo.